rclcpp/test/rclcpp/test_node.cpp

```diff
@@ -95,6 +95,74 @@ TEST_F(TestNode, construction_and_destruction) {
       (void)node;
     }, rclcpp::exceptions::InvalidNamespaceError);
   }
+
+  {
+    rclcpp::NodeOptions options;
+    ASSERT_NO_THROW(
+    {
+      const auto node = std::make_shared<rclcpp::Node>("my_node", "/ns", options);
+      EXPECT_EQ(RCL_ROS_TIME, node->get_clock()->get_clock_type());
+    });
+  }
+
+  {
+    rclcpp::NodeOptions options;
+    options.parameter_overrides(
+    {
+      {"use_sim_time", true},
+    });
+    ASSERT_NO_THROW(
+    {
+      const auto node = std::make_shared<rclcpp::Node>("my_node", "/ns", options);
+      EXPECT_EQ(RCL_ROS_TIME, node->get_clock()->get_clock_type());
+    });
+  }
+
+  {
+    rclcpp::NodeOptions options;
+    options.clock_type(RCL_SYSTEM_TIME);
+    ASSERT_NO_THROW(
+    {
+      const auto node = std::make_shared<rclcpp::Node>("my_node", "/ns", options);
+      EXPECT_EQ(RCL_SYSTEM_TIME, node->get_clock()->get_clock_type());
+    });
+  }
+
+  {
+    rclcpp::NodeOptions options;
+    options.parameter_overrides(
+    {
+      {"use_sim_time", true},
+    });
+    options.clock_type(RCL_SYSTEM_TIME);
+    ASSERT_THROW(
+      const auto node = std::make_shared<rclcpp::Node>(
+        "my_node", "/ns",
+        options), std::invalid_argument);
+  }
+
+  {
+    rclcpp::NodeOptions options;
+    options.clock_type(RCL_STEADY_TIME);
+    ASSERT_NO_THROW(
+    {
+      const auto node = std::make_shared<rclcpp::Node>("my_node", "/ns", options);
+      EXPECT_EQ(RCL_STEADY_TIME, node->get_clock()->get_clock_type());
+    });
+  }
+
+  {
+    rclcpp::NodeOptions options;
+    options.parameter_overrides(
+    {
+      {"use_sim_time", true},
+    });
+    options.clock_type(RCL_STEADY_TIME);
+    ASSERT_THROW(
+      const auto node = std::make_shared<rclcpp::Node>(
+        "my_node", "/ns",
+        options), std::invalid_argument);
+  }
```

Comment:
So user must make sure that if they want to use ROS simulation time, the Node::Option clock_type must be ROS time statically?

What about the user changes `use_sim_time` parameter to `true` when they want to use simulation time for the node afterward? i think this is coming back to @clalancette 's comment https://github.com/ros2/rclcpp/pull/1982#discussion_r947205947?
I do not think this is gonna work with simulation time if user enables `use_sim_time` later after node initialization.

Reply:
Since `on_parameter_event()` eventually calls `set_clock()` when `use_sim_time` is changed, it will cause an exception throw if the clock is not of `RCL_ROS_TIME` type. I believe it's the correct behavior as only `RCL_ROS_TIME` allows sim time.
source:
https://design.ros2.org/articles/clock_and_time.html#time-abstractions

Reply:
`clocks_state_.enable_ros_time`  in `attachNode` throwing an exception makes sense, but is it good to throw an exception in `on_parameter_event`? I think using `RCLCPP_ERROR` in the `on_parameter_event` might be better.
It seems that [the logic of throwing the exception](https://github.com/ros2/rclcpp/blob/bca2d75fc4caa532b38580d0135b48c79e111874/rclcpp/src/rclcpp/time_source.cpp#L148) should move up.

Reply:
> I believe it's the correct behavior as only RCL_ROS_TIME allows sim time.

i think this is correct. but related to previous @iuhilnehc-ynos comment, i do not think that is a good behavior.

> I think using RCLCPP_ERROR in the on_parameter_event might be better.

this could be one solution. but before that, probably it should not be able to change the parameter `use_sim_time` into `true` in this case? because we will throw the exception afterwards.
It would be better to reject the parameter change and let the user application know that `you cannot use simulation time, because node specifies to use system or monotonic clock type via option`?

Reply:
@fujitatomoya, I agree.
I added a new `on_set_parameters_callback` to reject `use_sim_time` parameter changes altogether if the clock_type is not compatible.

Reply:
@iuhilnehc-ynos could you review on this too to share your comment?